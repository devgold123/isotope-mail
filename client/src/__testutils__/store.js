import {createStore} from 'redux';
import rootReducer, {INITIAL_STATE} from '../reducers';

export const MOCK_STORE = createStore(rootReducer, INITIAL_STATE);
