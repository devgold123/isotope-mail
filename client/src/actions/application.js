import {ActionTypes} from './action-types';

export const backendRequest = () => ({type: ActionTypes.APPLICATION_BE_REQUEST});
export const backendRequestCompleted = () => ({type: ActionTypes.APPLICATION_BE_REQUEST_COMPLETED});
export const setUserCredentials = (userId, hash, credentials) =>
  ({type: ActionTypes.APPLICATION_USER_CREDENTIALS_SET, payload: {userId, hash, credentials}});
export const selectFolder = folder => ({type: ActionTypes.APPLICATION_FOLDER_SELECT, payload: folder});
export const selectMessage = message => ({type: ActionTypes.APPLICATION_MESSAGE_SELECT, payload: message});
/**
 * Refreshes the current selectedMessage if it's still the same (same selectedFolder and same selectedMessage)
 *
 * @param folder
 * @param message
 * @returns {{type: string, payload: {folder: *, message: *}}}
 */
export const refreshMessage = (folder, message) =>
  ({type: ActionTypes.APPLICATION_MESSAGE_REFRESH, payload: {folder, message}});
export const replaceMessageEmbeddedImages = (folder, message, attachment, blob) =>
  ({type: ActionTypes.APPLICATION_MESSAGE_REPLACE_IMAGE, payload: {folder, message, attachment, blob}});
export const setError = (type, value) => ({type: ActionTypes.APPLICATION_ERROR_SET, payload: {type, value}});

export const editMessage = message => ({type: ActionTypes.APPLICATION_MESSAGE_EDIT, payload: message});

export const sendMessage = message => ({type: ActionTypes.APPLICATION_OUTBOX_SEND, payload: message});
export const sendMessageUpdateProgress = progress =>
  ({type: ActionTypes.APPLICATION_OUTBOX_UPDATE_PROGRESS, payload: progress});
export const messageSent = () => ({type: ActionTypes.APPLICATION_OUTBOX_MESSAGE_SENT});
