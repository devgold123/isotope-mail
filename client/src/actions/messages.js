import {ActionTypes} from './action-types';

export const backendRequest = () => ({type: ActionTypes.MESSAGES_BE_REQUEST});
export const backendRequestCompleted = () => ({type: ActionTypes.MESSAGES_BE_REQUEST_COMPLETED});

export const setCache = cache => ({
  type: ActionTypes.MESSAGES_SET_CACHE,
  payload: cache
});

export const setFolderCache = (folder, messages) => ({
  type: ActionTypes.MESSAGES_SET_FOLDER_CACHE,
  payload: {folder, messages}
});

/**
 * Updates state folder's messages cache.
 *
 * @param folder to which the messages will be added
 * @param messages {Array} array of messages to add to the folder's cache
 * @returns {{type: string, payload: {folder: *, messages: *}}}
 */
export const updateCache = (folder, messages) => ({
  type: ActionTypes.MESSAGES_UPDATE_CACHE,
  payload: {folder, messages}
});

/**
 * Delete messages from state folder's messages cache.
 *
 * @param folder from which the messages will be deleted
 * @param messages {Array} array of messages to delete from the folder's cache
 * @returns {{type: string, payload: {folder: *, messages: *}}}
 */
export const deleteFromCache = (folder, messages) => ({
  type: ActionTypes.MESSAGES_DELETE_FROM_CACHE,
  payload: {folder, messages}
});

/**
 * Adds the provided message uid to the selected message list if selected or else removes it from the list
 *
 * @param message {Object}
 * @param selected {boolean}
 * @returns {{type: string, payload: {message: *, selected: *}}}
 */
export const setSelected = (message, selected) => ({
  type: ActionTypes.MESSAGES_SET_SELECTED,
  payload: {message, selected}
});

/**
 * Clears the list of selected message uids
 *
 * @returns {{type: string}}
 */
export const clearSelected = () => ({
  type: ActionTypes.MESSAGES_CLEAR_SELECTED
});
