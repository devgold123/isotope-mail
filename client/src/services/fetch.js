export const HttpHeaders = {
  ISOTOPE_CREDENTIALS: 'X-Isotope-Credentials',
  ISOTOPE_SALT: 'X-Isotope-Salt',
  CONTENT_TYPE: 'Content-Type'
};

export const HttpStatusFamilies = {
  INFORMATIONAL: 1,
  SUCCESSFUL: 2,
  REDIRECTION: 3,
  CLIENT_ERROR: 4,
  SERVER_ERROR: 5
};

/**
 * Object to store the different AbortController(s) that will be used in the service methods to fetch from the API backend.
 *
 * @type {Object}
 * @private
 */
export const abortControllerWrappers = {};

/**
 * Converts response Promise to json Promise or throws an error.
 *
 * @param {Response}response
 * @returns {PromiseLike<object>}
 */
export function toJson(response) {
  if (!response.ok) {
    throw Error(`${response.status}`);
  }
  return response.json();
}

/**
 * Triggers the abort method of a the specified AbortController
 *
 * @param abortController {AbortController}
 */
export function abortFetch(abortController) {
  if (abortController && abortController.abort) {
    abortController.abort();
  }
}

/**
 * Creates a new headers object including Isotope specific credentials headers for server authentication.
 *
 * @param {{}} credentials
 * @param {{}} [originalHeaders={}] originalHeaders
 * @returns {{}}
 */
export function credentialsHeaders(credentials, originalHeaders) {
  const newHeaders = Object.assign({}, originalHeaders ? originalHeaders : {});
  newHeaders[HttpHeaders.ISOTOPE_CREDENTIALS] = credentials.encrypted;
  newHeaders[HttpHeaders.ISOTOPE_SALT] = credentials.salt;
  return newHeaders;
}

/**
 * Returns true if the response the Successful HTTP status code range (2xx)
 *
 * @param HTTP status code of the response
 * @returns {boolean} true if status code belongs to HttpStatusFamilies.SUCCESSFUL family
 */
export function isSuccessful(status) {
  return Math.floor(status / 100) === HttpStatusFamilies.SUCCESSFUL;
}
