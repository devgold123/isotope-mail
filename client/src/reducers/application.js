import {INITIAL_STATE} from './';
import {ActionTypes} from '../actions/action-types';

const application = (state = INITIAL_STATE.application, action = {}) => {
  switch (action.type) {
    case ActionTypes.APPLICATION_BE_REQUEST:
      return {...state, activeRequests: state.activeRequests + 1};
    case ActionTypes.APPLICATION_BE_REQUEST_COMPLETED:
      return {...state, activeRequests: state.activeRequests > 0 ? state.activeRequests - 1 : 0};
    case ActionTypes.APPLICATION_USER_CREDENTIALS_SET:
      return {
        ...state, user: {
          ...state.user,
          id: action.payload.userId, hash: action.payload.hash, credentials: action.payload.credentials
        }
      };
    case ActionTypes.APPLICATION_FOLDER_SELECT:
      return {
        ...state,
        selectedFolderId: action.payload.folderId,
        selectedFolder: {...action.payload}
      };
    case ActionTypes.APPLICATION_MESSAGE_SELECT:
      return {...state, selectedMessage: {...action.payload}};
    case ActionTypes.APPLICATION_MESSAGE_REFRESH:
      if (state.selectedFolderId === action.payload.folder.folderId
        && state.selectedMessage.uid === action.payload.message.uid) {
        return {...state, selectedMessage: {...action.payload.message}};
      }
      return state;
    case ActionTypes.APPLICATION_MESSAGE_REPLACE_IMAGE: {
      if (state.selectedFolder.folderId === action.payload.folder.folderId
        && state.selectedMessage.uid === action.payload.message.uid) {
        const contentId = action.payload.attachment.contentId.replace(/[<>]/g, '');
        if (contentId.length > 0) {
          const objectUrl = URL.createObjectURL(action.payload.blob);
          // Multiple occurrence
          const parsedMessage = state.selectedMessage.content.replace(new RegExp(`cid:${contentId}`, 'g'), objectUrl);
          return {...state, selectedMessage: {...state.selectedMessage, content: parsedMessage}};
        }
      }
      return state;
    }
    case ActionTypes.APPLICATION_ERROR_SET:
      const errorsSetState = {...state};
      errorsSetState.errors = {...state.errors};
      errorsSetState.errors[action.payload.type] = action.payload.value;
      return errorsSetState;
    case ActionTypes.APPLICATION_MESSAGE_EDIT: {
      const newState = {...state};
      newState.newMessage = {...action.payload};
      return newState;
    }
    case ActionTypes.APPLICATION_OUTBOX_SEND: {
      const newState = {...state};
      newState.outbox = {message: action.payload, progress: 0, sent: false};
      return newState;
    }
    case ActionTypes.APPLICATION_OUTBOX_UPDATE_PROGRESS: {
      const newState = {...state};
      newState.outbox = {...newState.outbox, progress: action.payload};
      return newState;
    }
    case ActionTypes.APPLICATION_OUTBOX_SET_SENT: {
      const newState = {...state};
      newState.outbox = {...newState.outbox, sent: action.payload};
      return newState;
    }
    case ActionTypes.APPLICATION_OUTBOX_MESSAGE_SENT: {
      const newState = {...state};
      newState.outbox = null;
      return newState;
    }
    default:
      return state;
  }
};

export default application;
