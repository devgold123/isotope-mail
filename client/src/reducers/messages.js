import {INITIAL_STATE} from './';
import {ActionTypes} from '../actions/action-types';

const messages = (state = INITIAL_STATE.messages, action = {}) => {
  switch (action.type) {
    case ActionTypes.APPLICATION_FOLDER_RENAME_OK: {
      const newState = {...state};
      newState.cache[action.payload.newFolderId] = newState.cache[action.payload.oldFolderId];
      delete newState.cache[action.payload.oldFolderId];
      return newState;
    }
    case ActionTypes.MESSAGES_BE_REQUEST:
      return {...state, activeRequests: state.activeRequests + 1};
    case ActionTypes.MESSAGES_BE_REQUEST_COMPLETED:
      return {...state, activeRequests: state.activeRequests > 0 ? state.activeRequests - 1 : 0};
    case ActionTypes.MESSAGES_SET_CACHE:
      return {...state, cache: action.payload};
    case ActionTypes.MESSAGES_SET_FOLDER_CACHE: {
      const newState = {...state};
      newState.cache[action.payload.folder.folderId] =
        new Map(action.payload.messages.map(m => [m.uid, m]));
      return newState;
    }
    case ActionTypes.MESSAGES_UPDATE_CACHE: {
      const newUpdateState = {...state};
      if (newUpdateState.cache[action.payload.folder.folderId] instanceof Map === false) {
        newUpdateState.cache[action.payload.folder.folderId] = new Map();
      }
      action.payload.messages.forEach(m => newUpdateState.cache[action.payload.folder.folderId].set(m.uid, m));
      return newUpdateState;
    }
    case ActionTypes.MESSAGES_UPDATE_CACHE_IF_EXIST: {
      const newUpdateState = {...state};
      const cache = newUpdateState.cache[action.payload.folder.folderId];
      if (cache instanceof Map === true) {
        action.payload.messages.forEach(m => {
          if (cache.has(m.uid)) {
            cache.set(m.uid, m);
          }
        });
      }
      return newUpdateState;
    }
    case ActionTypes.MESSAGES_DELETE_FROM_CACHE: {
      const newUpdateState = {...state};
      if (newUpdateState.cache[action.payload.folder.folderId] instanceof Map === false) {
        return state;
      }
      action.payload.messages.forEach(m => newUpdateState.cache[action.payload.folder.folderId].delete(m.uid));
      return newUpdateState;
    }
    case ActionTypes.MESSAGES_SET_SELECTED: {
      const newUpdateState = {...state};
      newUpdateState.selected = [...state.selected];
      action.payload.messages.forEach(message => {
        const indexOfMessage = newUpdateState.selected.indexOf(message.uid);
        if (action.payload.selected && indexOfMessage < 0) {
          // Select Message
          newUpdateState.selected.push(message.uid);
        } else if (!action.payload.selected) {
          // Unselect message
          newUpdateState.selected = newUpdateState.selected.filter(uid => uid !== message.uid);
        }
      });
      return newUpdateState;
    }
    case ActionTypes.MESSAGES_CLEAR_SELECTED: {
      const newUpdateState = {...state, selected: []};
      return newUpdateState;
    }
    default:
      return state;
  }
};

export default messages;
