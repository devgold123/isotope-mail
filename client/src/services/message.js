import {
  backendRequest as aBackendRequest,
  backendRequestCompleted as aBackendRequestCompleted, replaceMessageEmbeddedImages
} from '../actions/application';
import {
  backendRequest,
  backendRequestCompleted,
  deleteFromCache,
  setFolderCache,
  updateCache
} from '../actions/messages';
import {refreshMessage} from '../actions/application';
import {updateFolder} from '../actions/folders';
import {abortControllerWrappers, abortFetch, credentialsHeaders, toJson} from './fetch';
import {processFolders} from './folder';
import {persistMessageCache} from './indexed-db';
import {KEY_HASH, KEY_USER_ID} from './state';

const _eventSourceWrappers = {};

function _closeEventSource(dispatch, es) {
  if (es && es.close) {
    es.close();
    if (!es.DISPATCHED) {
      dispatch(backendRequestCompleted());
      es.DISPATCHED = true;
      es.resolvePromise();
    }
  }
}

function _readContent(dispatch, credentials, folder, message, signal, attachment) {
  fetch(attachment._links.download.href, {
    method: 'GET',
    headers: credentialsHeaders(credentials),
    signal: signal
  })
    .then(response => response.blob())
    .then(blob => {
      dispatch(replaceMessageEmbeddedImages(folder, message, attachment, blob));
    });
}

/**
 *
 * @param dispatch
 * @param credentials
 * @param folder
 */
export async function resetFolderMessagesCache(dispatch, credentials, folder) {
  _closeEventSource(dispatch, _eventSourceWrappers.resetFolderMessagesCache);
  if (folder && folder._links) {
    const allMessages = [];
    // Prefer EventSourcePolyfill instead of EventSource to allow sending HTTP headers in all browsers
    const es = new window.EventSourcePolyfill(folder._links.messages.href,
      {
        headers: credentialsHeaders(credentials)
      });
    _eventSourceWrappers.resetFolderMessagesCache = es;
    dispatch(backendRequest());
    es.onmessage = e => {
      const messages = JSON.parse(e.data);
      allMessages.push(...messages);
      if (e.lastEventId === '1') {
        // This is the last batch
        dispatch(setFolderCache(folder, allMessages));
        _closeEventSource(dispatch, _eventSourceWrappers.resetFolderMessagesCache);
        // Manually persist newest version of message cache
        persistMessageCache(
          sessionStorage.getItem(KEY_USER_ID), sessionStorage.getItem(KEY_HASH), folder, [...allMessages]);
      } else {
        dispatch(updateCache(folder, messages));
      }
    };
    // Return a promise
    const eventSourceCompleted = new Promise(resolve => {
      // Store resolve function that will be called in _closeEventSource
      es.resolvePromise = resolve;
    });
    return eventSourceCompleted;
  }
  return null;
}

/**
 *
 * @param dispatch
 * @param credentials
 * @param folder
 * @param message {object}
 */
export function readMessage(dispatch, credentials, folder, message) {
  // Abort + signal
  if (message && message._links) {
    abortFetch(abortControllerWrappers.readMessageAbortController);
    abortControllerWrappers.readMessageAbortController = new AbortController();
    const signal = abortControllerWrappers.readMessageAbortController.signal;

    dispatch(aBackendRequest());
    fetch(message._links.self.href, {
      method: 'GET',
      headers: credentialsHeaders(credentials),
      signal: signal
    })
      .then(response => {
        dispatch(aBackendRequestCompleted());
        return response;
      })
      .then(toJson)
      .then(completeMessage => {
        dispatch(refreshMessage(folder, completeMessage));
        // Update folder with freshest information
        dispatch(updateFolder(processFolders([completeMessage.folder])[0]));
        // Update folder cache with message marked as read (don't store content in cache)
        const messageWithNoContent = {...completeMessage};
        delete messageWithNoContent.content;
        dispatch(updateCache(folder, [messageWithNoContent]));
        // Read message's embedded images
        completeMessage.attachments
          .filter(a => a.contentId && a.contentId.length > 0)
          .filter(a => a.contentType.toLowerCase().indexOf('image/') === 0)
          .forEach(a => _readContent(dispatch, credentials, folder, message, signal, a));
      }).catch(() => dispatch(aBackendRequestCompleted()));
  }
}

export function downloadAttachment(credentials, attachment) {
  const fetch$ = fetch(attachment._links.download.href, {
    method: 'GET',
    headers: credentialsHeaders(credentials)
  }).then(response => response.blob());
  fetch$
    .then(blob => {
      if (navigator.msSaveBlob) {
        navigator.msSaveBlob(blob, attachment.fileName);
      } else {
        const url = URL.createObjectURL(blob);
        const tempLink = document.createElement('a');
        tempLink.href = url;
        tempLink.download = attachment.fileName;
        document.body.appendChild(tempLink);
        // tempLink.click();
        tempLink.dispatchEvent(new MouseEvent('click', {bubbles: true, cancelable: true, view: window}));
        document.body.removeChild(tempLink);
        URL.revokeObjectURL(url);
      }
    });
  return fetch$;
}

export function moveMessage(dispatch, credentials, fromFolder, toFolder, message) {
  // Abort any operations that can affect operation result
  _closeEventSource(dispatch, _eventSourceWrappers.resetFolderMessagesCache);
  abortFetch(abortControllerWrappers.getFoldersAbortController);

  // Calculate new fromFolder values
  const fromFolderUpdated = {
    ...fromFolder,
    messageCount: fromFolder.messageCount - 1,
    unreadMessageCount: fromFolder.unreadMessageCount - (message.seen ? 0 : 1),
    newMessageCount: fromFolder.newMessageCount - (message.recent ? 1 : 0)
  };

  // Update state with expected response from server
  dispatch(deleteFromCache(fromFolder, [message]));
  dispatch(updateFolder(fromFolderUpdated));
  fetch(message._links.move.href.replace('{toFolderId}', toFolder.folderId), {
    method: 'PUT',
    headers: credentialsHeaders(credentials)
  })
    .then(toJson)
    .then(newMessages => {
      if (Array.isArray(newMessages)) {
        dispatch(updateCache(toFolder, newMessages));
        newMessages.forEach(m => dispatch(updateFolder(processFolders([m.folder])[0])));
      }
    })
    .catch(() => {
      // Rollback state from dispatched expected responses
      dispatch(updateCache(fromFolder, [message]));
      dispatch(updateFolder(fromFolder));
    });
}

export function setMessageSeen(dispatch, credentials, folder, message, seen) {
  // Abort any operations that can affect operation result
  _closeEventSource(dispatch, _eventSourceWrappers.resetFolderMessagesCache);
  abortFetch(abortControllerWrappers.getFoldersAbortController);

  // Update state with expected response from server
  const expectedUpdatedMessage = {...message, seen: seen};
  dispatch(updateCache(folder, [expectedUpdatedMessage]));
  const expectedUpdatedFolder = {...folder, unreadMessageCount: folder.unreadMessageCount + 1};
  dispatch(updateFolder(expectedUpdatedFolder));

  fetch(message._links.seen.href, {
    method: 'PUT',
    headers: credentialsHeaders(credentials, { 'Content-Type': 'application/json'}),
    body: JSON.stringify(seen)
  })
    .then(toJson)
    .then(updatedMessage => {
      dispatch(updateCache(folder, [updatedMessage]));
      dispatch(updateFolder(processFolders([updatedMessage.folder])[0]));
    })
    .catch(() => {
      // Rollback state from dispatched expected responses
      dispatch(updateCache(folder, [message]));
      dispatch(updateFolder(folder));
    });
}
