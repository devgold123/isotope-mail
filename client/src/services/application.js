import sjcl from 'sjcl';
import {URLS} from './url';
import {
  backendRequest,
  backendRequestCompleted,
  editMessage,
  selectFolder,
  setUserCredentials
} from '../actions/application';
import {toJson} from './fetch';
import {FolderTypes, getFolders} from './folder';
import i18n from './i18n';
import {recoverState} from './indexed-db';
import {setFolders} from '../actions/folders';
import {setCache} from '../actions/messages';
import {resetFolderMessagesCache} from './message';
import sanitize from './sanitize';

export const DEFAULT_IMAP_PORT = 993;
export const DEFAULT_IMAP_SSL = true;
export const DEFAULT_SMTP_PORT = 465;
export const DEFAULT_SMTP_SSL = true;

/**
 * @typedef {Object} Credentials
 * @property {string} serverHost - IMAP host
 * @property {string} serverPort - IMAP port
 * @property {string} user - User
 * @property {string} password - Password
 */
/**
 * Performs a login to the IMAP server using the API.
 *
 * If the server validates the credentials, previous state is reloaded (if exists) from IndexedDB.
 *
 * If previous state exists, message cache, folder list, and selected folder will be pre-populated with existing values.
 * An initial call to the backend will be made to refresh the list of messages of the selected folder.
 *
 * If a previous state doesn't exist, an initial load of the first level folders will be requested to the server.
 *
 * @param dispatch {Dispatch & function}
 * @param credentials {Credentials}
 * @returns {Promise<void>}
 */
export async function login(dispatch, credentials) {
  dispatch(backendRequest());
  const url = URLS.LOGIN;
  // Will be used as the key in the IndexedDB
  const userId = sjcl.codec.hex.fromBits(
    sjcl.hash.sha256.hash(`${credentials.serverHost}|${credentials.user}`));
  // Will be used as the encryption password to store state in the IndexedDB
  const hash = sjcl.codec.hex.fromBits(
    sjcl.hash.sha256.hash(`${credentials.serverHost}|${credentials.user}|${credentials.password}`));

  const response = await fetch(url, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify(credentials)
  });
  dispatch(backendRequestCompleted());
  if (response.ok) {
    const validatedCredentials = await toJson(response);
    dispatch(setUserCredentials(userId, hash, validatedCredentials));
    // Reload data from indexedDb
    const recoveredState = await recoverState(userId, hash);
    if (recoveredState !== null) {
      await dispatch(setCache(recoveredState.messages.cache));
      await dispatch(setFolders(recoveredState.folders.items));
      dispatch(selectFolder(recoveredState.application.selectedFolder));
    } else {
      // Retrieve first level folders to show something ASAP
      const folders = await getFolders(dispatch, validatedCredentials, false);
      // Retrieve and select INBOX folder so that user has something in the screen
      const inbox = folders.find(f => f.type === FolderTypes.INBOX);
      if (inbox) {
        dispatch(selectFolder(inbox));
        resetFolderMessagesCache(dispatch, validatedCredentials, inbox, null);
      }
    }
  }
}

export function editNewMessage(dispatch) {
  dispatch(editMessage({to: [], cc: [], bcc: [], subject: '', content: ''}));
}

export function replyMessage(dispatch, originalMessage) {
  const recipients = [...originalMessage.recipients];
  const recipientMapper = r => r.address;
  const to = recipients.filter(r => r.type === 'To').map(recipientMapper).concat(originalMessage.from);
  const cc = recipients.filter(r => r.type === 'Cc').map(recipientMapper);
  const bcc = recipients.filter(r => r.type === 'Bcc').map(recipientMapper);
  const subject = `Re: ${originalMessage.subject}`;
  const formattedDate = new Date(originalMessage.receivedDate).toLocaleString(navigator.language, {
    year: 'numeric', month: '2-digit', day: '2-digit', hour: '2-digit', minute: '2-digit', second: '2-digit'
  });

  const t = i18n.t.bind(i18n);
  const content = `
    <p></p>
    <hr/>
    <p>
      <b>${t('replyAction.From')}:</b> ${originalMessage.from.join(', ')}<br/>
      <b>${t('replyAction.Date')}:</b> ${formattedDate}<br/>
      <b>${t('replyAction.Subject')}:</b> ${originalMessage.subject}<br/>
    </p>
    <br/>
    ${sanitize.sanitize(originalMessage.content)}
  `;

  dispatch(editMessage({to, cc, bcc, subject, content
  }));
}
