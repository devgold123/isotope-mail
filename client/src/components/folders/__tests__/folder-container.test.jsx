import React from 'react';
import {shallow} from 'enzyme/build/index';
import {MOCK_STORE} from '../../../__testutils__/store';
import * as messageService from '../../../services/message';
import * as applicationActions from '../../../actions/application';
import * as messageActions from '../../../actions/messages';
import ConnectedFolderContainer, {FolderContainer} from '../folder-container';

describe('FolderContainer component test suite', () => {
  test('Snapshot render, should render FolderContainer', () => {
    const {activeRequests, folderList, selectedFolder} = {
      activeRequests: 0, folderList: [{folderList: []}], selectedFolder: {folderId: '1337-3'}};
    const folderContainer = shallow(
      <FolderContainer activeRequests={activeRequests} folderList={folderList}
        selectedFolder={selectedFolder}/>);
    expect(folderContainer).toMatchSnapshot();
  });
  test('selectFolder, dispatch actions triggered', () => {
    // Given
    const mockAction = {type: 'MOCK', action: 'MOCK'}
    applicationActions.selectFolder = jest.fn(() => (mockAction));
    applicationActions.selectMessage = jest.fn(() => (mockAction));
    messageActions.clearSelected = jest.fn(() => (mockAction));
    messageService.resetFolderMessagesCache = jest.fn();
    const {activeRequests, folderList, selectedFolder} = {
      activeRequests: 0, folderList: [{folderList: []}], selectedFolder: {folderId: '1337-3'}};
    const folderContainer = shallow(
      <ConnectedFolderContainer store={MOCK_STORE} activeRequests={activeRequests} folderList={folderList}
        selectedFolder={selectedFolder}/>);

    // When
    folderContainer.props().selectFolder({});

    // Then
    expect(applicationActions.selectFolder).toHaveBeenCalledTimes(1);
    expect(applicationActions.selectMessage).toHaveBeenCalledTimes(1);
    expect(messageActions.clearSelected).toHaveBeenCalledTimes(1);
    expect(messageService.resetFolderMessagesCache).toHaveBeenCalledTimes(1);
  });
  test('moveMessages, service called', () => {
    // Given
    messageService.moveMessages = jest.fn();
    const {activeRequests, folderList, selectedFolder} = {
      activeRequests: 0, folderList: [{folderList: []}], selectedFolder: {folderId: '1337-3'}};
    const folderContainer = shallow(
      <ConnectedFolderContainer store={MOCK_STORE} activeRequests={activeRequests} folderList={folderList}
        selectedFolder={selectedFolder}/>);

    // When
    folderContainer.props().moveMessages();

    // Then
    expect(messageService.moveMessages).toHaveBeenCalledTimes(1);
  });
});
