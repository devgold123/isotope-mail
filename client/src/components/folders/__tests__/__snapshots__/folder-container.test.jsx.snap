// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FolderContainer component test suite Snapshot render, should render FolderContainer 1`] = `
<nav
  className="mdc-list"
>
  <Spinner
    canvasClassName="spinnerCanvas"
    className=""
    pathClassName=""
    visible={false}
  />
  <FolderList
    folderList={
      Array [
        Object {
          "folderList": Array [],
        },
      ]
    }
    onClickFolder={[Function]}
    onDropMessage={[Function]}
    onRenameFolder={[Function]}
    selectedFolder={
      Object {
        "folderId": "1337-3",
      }
    }
  />
  <Connect(Translate(FolderRenameDialog)) />
</nav>
`;
